update corresponding authors (#1)
ChenZhao44/ChenZhao44.github.io: _posts/2020-06-21-Quantum circuits simplification and ZX-calculus.md

---
layout: post
title: Quantum circuits simplification and ZX-calculus
description: Implement ZX-calculus with Julia.
---

> Quantum computing uses the principles of quantum mechanics for computing. To describe quantum algorithms, we often use quantum circuit model which is the analog of classical logic circuit model. Quantum circuits are consist of a set of basic quantum gates. For example, Pauli X, Y, Z gate, Hadamard gate, T gate, CNOT gate. And circuits will apply on qubits which are initiallized as input. Finally, one measure the qubits to get results.

Reviewer: ```suggestion
Quantum computing uses the principles of quantum mechanics for computing. To describe quantum algorithms, we often use the quantum circuit model which is the analog of the classical logic circuit model. Quantum circuits are consist of a set of basic quantum gates such as Pauli X, Y, Z gate, Hadamard gate, T gate, CNOT gate.
```

> In general, there will be lots of equivilant quantum circuits representing the same operation. At the view point of quantum hardwares, one wish to find the most elegant circuit for saving hardware resources. Unfortunatelly, finding the most simple circuit is a very difficult problem in computational complexity, as checking the equavilance between two quantum circuits is coQMA-hard (quantum analog of coNP). Despite this, we still have some heuristic efficient algorithms for quantum circuits simplification. In this blog post, I will introduce a powerful tool for this problem, ZX-calculus.

Reviewer: ```suggestion
In general, there will be lots of equivalent quantum circuits representing the same operation. At the viewpoint of quantum hardware, one wishes to find the circuit that minimizing the usage of hardware resources. Unfortunately, finding the most desired circuit is a very hard problem in computational complexity, as checking the equivalence between two quantum circuits is coQMA-hard (quantum analog of coNP). Despite this, we still have some heuristic efficient algorithms for quantum circuits simplification. In this blog post, I will introduce a powerful tool for this problem, ZX-calculus.
```

A quantum circuit is a graph representing tensor products and matrices products. For example, this is the circuits representing 
$$
\mathrm{CNOT}(2,3)\cdot\mathrm{CNOT}(1,2)\cdot(H\otimes I\otimes I).
$$
![](\assets\blog_res\ZX\cir1.png)

> Also, quantum circuits can be regarded as a special type of tensor networks. In ZX-calculus, we introduce two kinds of tensors, Z-spiders and X-spiders. Tensor networks which consist of these tensors are called ZX-diagrams. As all basic quantum gates can be represented by ZX-diagrams, we can transform all quantum circuits to ZX-diagrams. 

Reviewer: ```suggestion
Also, quantum circuits can be regarded as a special type of tensor network. In ZX-calculus, we introduce two kinds of tensors, Z-spiders, and X-spiders. Tensor networks that consist of these tensors are called ZX-diagrams. As all basic quantum gates can be represented by ZX-diagrams, we can transform all quantum circuits to ZX-diagrams. 
```

![](\assets\blog_res\ZX\zxd1.svg)
Moreover, by the definition of Z-spiders and X-spiders, there are basic equivilant rules with which one can simplify ZX-diagrams. 
![rules](\assets\blog_res\ZX\rules.png)

After simplifying ZX-diagrams, we should transform them back to circuits. In the paper [arXiv:1902.03178](https://arxiv.org/abs/1902.03178), they developed algorithms for the above scheme. We will use Julia to implement this scheme for simplifying quantum circuits.

> ## Data structs

Reviewer: ```suggestion
## Data structures
```

> To implement these algorithms, we need to implement data structs for storing ZX-diagrams and rules how ZX-diagrams change. A ZX-diagram can be regarded as a multigraph with extra information on its vertices. Each vertex is one of Z-spider, X-spider or H-box. In the original ZX-diagrams, open edges are allowed. For simplicity, we added 2 special kinds of vertices, input boundary and output boundary, for recording these edges. We use the adjancency list representation for storing multigraphs. And adjancency list is stored as `Dict` whose keys are the ids of vertices. 

Reviewer: ```suggestion
To implement these algorithms, we need to implement data structures for storing ZX-diagrams and rules on how ZX-diagrams change. A ZX-diagram can be regarded as a multigraph with extra information on its vertices. Each vertex is one of Z-spider, X-spider, or H-box. In the original ZX-diagrams, open edges are allowed. For simplicity, we added 2 special kinds of vertices, input boundary, and output boundary, for recording these edges. We use the adjacency list representation for storing multigraphs. An adjacency list is stored as `Dict` whose keys are the ids of vertices. 
```

```julia
struct Multigraph{T<:Integer} <: AbstractMultigraph{T}
    adjlist::Dict{T, Vector{T}}
end
```
Because we need to rewrite multigraphs with ZX-calculus rules, it will be more convenient to fix vertices ids. So we use `Dict` instead of `Array`. There are phases for Z-spiders and X-spiders. We need another Dict for storing these phases. The data struct of ZX-diagram is similar as follow:
```julia
struct ZXDiagram{T<:Integer, P} <: AbstractZXDiagram{T, P}
    mg::Multigraph{T}
    st::Dict{T, SpiderType.SType}
    ps::Dict{T, P}
    ...
end
```

> In the paper, they proposed graph-like ZX-diagrams. Graph-like ZX-diagrams have only Z-spiders and have different types of edges, Hadamard edges and non-Hadamard edges. We can also use multigraphs for representing graph-like ZX-diagrams. We can use different edge multiplicities for representing different edge types. We define `ZXGraph` for graph-like ZX-diagrams:

Reviewer: ```suggestion
In the paper, they proposed graph-like ZX-diagrams. Graph-like ZX-diagrams have only Z-spiders and have different types of edges, Hadamard edges, and non-Hadamard edges. We can also use multigraphs for representing graph-like ZX-diagrams. We can use different edge multiplicities for representing different edge types. We define `ZXGraph` for graph-like ZX-diagrams:
```

end
```
And `rewrite!` will try to use a rule to rewrite a ZX-diagram on the all matched vertices. Since it is possible that some matched vertices become unsatisfied with the corresponding rule after rewriting some vertices. We need to check whether a rule is still available on matched vertices. 
```julia
check_rule(r::AbstractRule, zxd::AbstractZXDiagram{T, P}, vs::Vector{T}) where {T, P}
```

For simpilifying ZX-diagram with a rule, we defined `simplify!` which is like
```julia
function simplify!(r::AbstractRule, zxd::ZXDiagram)
    matches = match(r, zxd)
    while length(matches) > 0
        rewrite!(r, zxd, matches)
        matches = match(r, zxd)
    end
    return zxd
end
```

In the simplification scheme of [arXiv:1902.03178](https://arxiv.org/abs/1902.03178), we need to convert a ZX-diagram to a graph-like ZX-diagram with rules i1, i2, h and f. Then simplify the graph-like ZX-diagram with local complementary rule and pivoting rule. We can accomplish these steps with above functions. Simplified graph-like ZX-diagrams will be only small skeletons comparing to original large circuits. The only thing remained is extracting circuits from ZX-diagrams.

## Circuit extraction

> This is the most complex part of the simplification scheme. For more detail, one can read the original paper. I will only explain the circuit extraction algorithm briefly.

Reviewer: ```suggestion
This is the most complicated part of the simplification scheme. For more detail, one can read the original paper. I will only explain the circuit extraction algorithm briefly.
```

A quantum circuit is a graph representing tensor products and matrices products. For example, this is the circuits representing 
$$
\mathrm{CNOT}(2,3)\cdot\mathrm{CNOT}(1,2)\cdot(H\otimes I\otimes I).
$$
![](\assets\blog_res\ZX\cir1.png)

Also, quantum circuits can be regarded as a special type of tensor network. In ZX-calculus, we introduce two kinds of tensors, Z-spiders, and X-spiders. Tensor networks that consist of these tensors are called ZX-diagrams. As all basic quantum gates can be represented by ZX-diagrams, we can transform all quantum circuits to ZX-diagrams. 
![](\assets\blog_res\ZX\zxd1.svg)
Moreover, by the definition of Z-spiders and X-spiders, there are basic equivalent rules with which one can simplify ZX-diagrams. 
![rules](\assets\blog_res\ZX\rules.png)
After simplifying ZX-diagrams, we should transform them back to circuits. In the paper [^3], they developed algorithms for the above scheme. 

During GSoC 2020, I'm going to achieve a Julia implementation of ZX-calculus. In the following sections, I will explain the reasons and methods briefly for this project.

## Proposal for this GSoC project

> The main proposal of this project is to implement ZX-calculus and its related algorithm in Julia language. And we will release a Julia package `ZXCalculus.jl`. There exists a Python implementation of ZX-calculus, [`PyZX`](https://github.com/Quantomatic/pyzx). Like `PyZX`, `ZXCalculus.jl` will provide APIs for rewriting ZX-diagrams, extracting circuits from ZX-diagrams, simplifying circuits. Visualization of ZX-diagrams will be available in `ZXCalculus.jl` too. Most functions of `PyZX` will be implemented in `ZXCalculus.jl`. Besides these, we will integrate `ZXCalculus.jl` in the quantum compiler [`YaoLang.jl`](https://github.com/QuantumBFS/YaoLang.jl) and implement a compiling level circuits simplification engine. In `YaoLang.jl`, one can define hybrid quantum programs with classical information (e.g. classical parameters, classical control flows). These quantum programs will be transformed into intermediate representations, `YaoIR`s, which contain quantum circuits. These circuits will be simplified with `ZXCalculus.jl`. Then `YaoIR`s will be compiled to backend instructions, like QASM or the simulator `Yao.jl`.

Reviewer: ```suggestion
The main proposal of this project is to implement ZX-calculus and its related algorithm in Julia language. And we will release a Julia package `ZXCalculus.jl`. There exists a Python implementation of ZX-calculus, [`PyZX`](https://github.com/Quantomatic/pyzx). Like `PyZX`, `ZXCalculus.jl` will provide APIs for rewriting ZX-diagrams, extracting circuits from ZX-diagrams, simplifying circuits. Visualization of ZX-diagrams will be available in `ZXCalculus.jl` too. Most functions of `PyZX` will be implemented in `ZXCalculus.jl`. Besides these, we will integrate `ZXCalculus.jl` in the quantum compiler [`YaoLang.jl`](https://github.com/QuantumBFS/YaoLang.jl) and implement a compilation-level circuit simplification engine.

In `YaoLang.jl`, one can define hybrid quantum programs with classical information (e.g. classical parameters, classical control flows). These quantum programs will be transformed into an intermediate representation (the Yao IR), which contains quantum circuits along with SSA (static single assignment) form of the classical program. These circuits will be simplified with `ZXCalculus.jl`. Then `YaoIR`s will be compiled to backend instructions, like QASM or the simulator instructions of Yao.
```

> For the efficiency of `YaoLang.jl`, a pure Julia implementation of ZX-calculus is necessary. And this is the start of `ZX-calculus.jl`.

Reviewer: ```suggestion
To achieve the best performance in the compilation and enable further customization of the simplification procedure, a pure Julia implementation of ZX-calculus is necessary. 
```
